src/Type/Php/RegexShapeMatcher.php

```diff
@@ -0,0 +1,176 @@
+use function is_string;
+use function preg_match;
+use function preg_replace;
+use function str_contains;
+use const PREG_OFFSET_CAPTURE;
+use const PREG_UNMATCHED_AS_NULL;
+
+final class RegexShapeMatcher
+{
+
+	/**
+	 * @param int-mask<PREG_OFFSET_CAPTURE|PREG_UNMATCHED_AS_NULL>|null $flags
+	 */
+	public function matchType(string $regex, ?int $flags, TypeSpecifierContext $context): ?Type
+	{
+		// add one capturing group to the end so all capture group keys
+		// are present in the $matches
+		// see https://3v4l.org/sOXbn, https://3v4l.org/3SdDM
+		$captureGroupsRegex = preg_replace('~.[a-z\s]*$~i', '|(?<phpstanNamedCaptureGroupLast>)$0', $regex);
+
+		if (
+			$captureGroupsRegex === null
+			|| @preg_match($captureGroupsRegex, '', $matches, PREG_UNMATCHED_AS_NULL) === false
+		) {
+			return null;
+		}
+		unset($matches[array_key_last($matches)]);
+		unset($matches['phpstanNamedCaptureGroupLast']);
+
+		// XXX hoa/regex does not support named capturing groups
+		$remainingNonOptionalGroupCount = $this->countNonOptionalGroups($regex);
+		if ($remainingNonOptionalGroupCount === null) {
+			// regex could not be parsed by Hoa/Regex
+			return null;
+		}
+
+		$builder = ConstantArrayTypeBuilder::createEmpty();
+		$valueType = $this->getValueType($flags ?? 0);
+
+		foreach (array_keys($matches) as $key) {
+			if ($key === 0) {
+				// first item in matches contains the overall match.
+				$builder->setOffsetValueType(
+					$this->getKeyType($key),
+					TypeCombinator::removeNull($valueType),
+					!$context->true(),
+				);
+
+				continue;
+			}
+
+			if (!$context->true()) {
+				$optional = true;
+			} else {
+				$optional = $remainingNonOptionalGroupCount <= 0;
+
+				if (is_int($key)) {
+					$remainingNonOptionalGroupCount--;
+				}
+			}
+
+			$builder->setOffsetValueType(
+				$this->getKeyType($key),
+				$valueType,
+				$optional,
+			);
+		}
+
+		return $builder->getArray();
+	}
+
+	private function getKeyType(int|string $key): Type
+	{
+		if (is_string($key)) {
+			return new ConstantStringType($key);
+		}
+
+		return new ConstantIntegerType($key);
+	}
+
+	private function getValueType(int $flags): Type
+	{
+		$valueType = new StringType();
+		$offsetType = IntegerRangeType::fromInterval(0, null);
+		if (($flags & PREG_UNMATCHED_AS_NULL) !== 0) {
+			$valueType = TypeCombinator::addNull($valueType);
+			// unmatched groups return -1 as offset
+			$offsetType = IntegerRangeType::fromInterval(-1, null);
+		}
+
+		if (($flags & PREG_OFFSET_CAPTURE) !== 0) {
+			$builder = ConstantArrayTypeBuilder::createEmpty();
+
+			$builder->setOffsetValueType(
+				new ConstantIntegerType(0),
+				$valueType,
+			);
+			$builder->setOffsetValueType(
+				new ConstantIntegerType(1),
+				$offsetType,
+			);
+
+			return $builder->getArray();
+		}
+
+		return $valueType;
+	}
+
+	private function countNonOptionalGroups(string $regex): ?int
+	{
+		/** @throws void */
+		$parser = Llk::load(new Read('hoa://Library/Regex/Grammar.pp'));
```

Comment:
should this be moved into a external class, so the grammar is not read over and over?

Reply:
Load once into static property?